src/config/pragmas/src-dirs.js

```diff
@@ -1,22 +1,37 @@
 let lambdaPragmas = require('../../defaults/lambda-pragmas')
 
-module.exports = function collectSourceDirs ({ pragmas }) {
+module.exports = function collectSourceDirs ({ arc, inventory, pragmas }) {
   let lambdaSrcDirs = []
   let unsortedBySrcDir = {}
+  function registerLambda (item, pragma) {
+    lambdaSrcDirs.push(item.src)
+    unsortedBySrcDir[item.src] = unsortedBySrcDir[item.src]
+      ? [ unsortedBySrcDir[item.src], { ...item, pragma } ] // Multiple Lambdae may map to a single dir
+      : { ...item, pragma }
+  }
   Object.entries(pragmas).forEach(([ pragma, values ]) => {
-    let mayHaveSrcDirs = lambdaPragmas.some(p => p === pragma)
-    if (mayHaveSrcDirs && Array.isArray(values)) {
-      pragmas[pragma].forEach(item => {
-        if (item.arcStaticAssetProxy === true) return // Special exception for ASAP
-        else if (typeof item.src === 'string') {
-          lambdaSrcDirs.push(item.src)
-          unsortedBySrcDir[item.src] = unsortedBySrcDir[item.src]
-            ? [ unsortedBySrcDir[item.src], { ...item, pragma } ] // Multiple Lambdae may map to a single dir
-            : { ...item, pragma }
+    if (pragma == 'plugins') {
+      Object.values(pragmas[pragma]).forEach(pluginModule => {
+        if (pluginModule && pluginModule.pluginFunctions) {
+          pluginModule.pluginFunctions(arc, { inv: inventory }).forEach(lambda => registerLambda(lambda, 'plugin'))
```

Comment:
This may be the key bit of feedback from this review: I suspect this may need a lot of validation. This is the first time we're making Lambdas pluggable, so we need to ensure they have the full Lambda Inventory signature because we have lots of consumers down the line relying on that standard signature.

Off the top of my head, that means we should ensure `name`, `config`, `src`, `handlerFile`, `handlerFunction`, and `configFile`, but more specifically see the ops in https://github.com/architect/inventory/blob/master/src/config/pragmas/populate-lambda/index.js (namely L66-L74).

Reply:
Makes sense. Will implement that (and will add throwing an exception in the case validation fails)! Thanks for catching that!

Reply:
Taking another look at this, I'm not sure the properties you point out are needed in this situation. Specifically, what's happening in the `src-dirs` Inventory module is only registering Lambda source paths - whether for core pragma Lambdae or plugin-authored ones.

As far as I can tell, the only place where `config` is used is during `package`ing when assembling the CloudFormation JSON. Per my RFC, setting those properties appropriately is handled entirely by the plugin author in a separate plugin interface method (the `package` method - which is basically the current method exposed by macros), and furthermore, is made transparent to the author with the use of the helper `createLambdaJSON` function that `package` will expose. For details on how this will work, you can check the "Example createLambdaJSON Usage" section of the [RFC](https://github.com/architect/architect/issues/1062) and the [related PR for adding `createLambdaJSON`](https://github.com/architect/package/pull/101).

`handlerFile`, `name`, and `handlerFunction` is not used at all, though something like `handlerFile` [_is_ used in `create` (but is hard-coded there)](https://github.com/architect/create/blob/master/src/lambda/index.js#L24). `package` doesn't use the `handlerFile` property when creating CloudFormation definitions for Lambdae, it only uses `src`.

Additionally, the `inv.plugins` map is a property focussed on exposing the plugin module - nothing more. Not all plugins will be creating new Lambdae.

If you wish, we could populate _another_ inventory property (say, `inv.pluginLambdas`) that would contain all of these properties for consistency with the core arc pragma lambdae. That said and as far as I can tell, these would not be used elsewhere within arc. I think this could be implemented by using Inventory's `read` module for the `config`, and we would use `config.handler` (which always returns the string `index.handler`) and parse out `handlerFile` and `handlerFunction` from these.

Reply:
For the purposes of this first pass at plugins support, I am going to resolve this particular comment and move forward.

I think we can revisit this issue in a future iteration of plugins support, in particular a new inventory property that tracks any plugin-created lambdas (i.e. `inv.pluginLambdas`).

Reply:
Unresolving this! Turns out you were prophetic here @ryanblock. Steaming ahead without addressing your original feedback is making code in other areas that need to resolve the logical ID of Lambdas (`logs` and direct `deploy` in particular) pretty ugly.

I will address this by storing the `require`d plugin modules under `inv._project.plugins` whereas `inv.plugins` will contain an array of objects representing plugin-exported Lambdas compatible with the arc lambda 'signature' object (containing properties like `config`, `name`, etc.)

Reply:
OK I am resolving this with a clear conscience this time! Feedback addressed and appreciated!